File: leetcode/560.和为k的子数组.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode.cn id=560 lang=java
 *
 * [560] 和为K的子数组
 */

// @lc code=start
class Solution {
    public int subarraySum(int[] nums, int k) {
        int res = 0;
        int numSum = 0;
        Map<Integer, Integer> sumMap = new HashMap<>();
        sumMap.put(0, 1);
        for (int i = 0; i < nums.length; i++) {
            numSum += nums[i];
            if (sumMap.containsKey(numSum - k)) {
                res += sumMap.get(numSum - k);
            }
            sumMap.put(numSum, sumMap.getOrDefault(numSum, 0) + 1);
        }
        return res;
    }

    // public static void main(String[] args) {
    // int[] nums = new int[] {1, 1, 1};
    // Solution solution = new Solution();
    // System.out.println(solution.subarraySum(nums, 2));
    // }
}
// @lc code=end

File: leetcode/560.和为k的子数组.py
#
# @lc app=leetcode.cn id=560 lang=python3
#
# [560] 和为K的子数组
#

# @lc code=start
from typing import List


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        # 前缀和
        result, num_sum = 0, 0
        sum_list = {0: 1}
        for num in nums:
            num_sum += num
            if num_sum - k in sum_list:
                result += sum_list[num_sum - k]
            sum_list[num_sum] = sum_list.get(num_sum, 0) + 1
        return result


print(Solution().subarraySum(nums=[1, 1, 1], k=2))
# @lc code=end
